### PR TITLE
make handle_reactions() faster by reducing redundancy in the reactions lists

### PR DESCRIPTION
**Pull request by xXxEMBERGHOSTSQUADDIExXx**
## What this does
the global chemical reactions list is a "list of lists" made of lists of reactions corresponding to a given reactant, let's call it the "key" reactant. this helps avoid checking a long list of all reactions, also by having a "key reactant" for each reaction, that also reduces redundancy in terms of:
eg., a reaction that requires oxygen and water, doesn't need to be checked for oxygen, if it's already checked that it doesn't contain water

reactions can also have lists of interchangable reactants listed as individual reactants. so you can have something that requires any of a given list of interchangable reagents. thats how botany chems can be substituted for chemlab chems, etc.

prior to #33390, when encountering a "list reactant", to avoid redundancy, the reaction would only be added to the reaction list of the first interchangable reactant in the "list reactant", but that caused a bug where botany chems etc. couldnt be used as intended. #33390 fixed this but also added some redundancy again in the case of "list reactants", eg:

`#define ACIDS list(SACID, PACID, FORMIC_ACID, PACID, PHENOL, ACIDSPIT, ACIDTEA)`
if ACIDS were the first required reactant for a reaction, the reaction would be added to 7 different lists
but, if the reaction also required water, we could alternatively add it to the water list once instead

this pr minimizes this redundancy by trying to avoid using "list reactants" as the key reactant where possible, and if a reaction only has "list reactants" then it tries to use the shortest list (this would now be the only source of duplicate reactions in the global list)

testing by running handle_reactions() 100000x on the "add 1u of everything in the chem dispenser but water" solution... it goes from 7.7 sec to 5.0 sec, a 35% reduction.

## Why it's good
faster handle_reactions(), increase performance

### CODE REVIEW

File: code/modules/reagents/Chemistry-Holder.dm [hunk 1]
```diff
@@ -44,19 +44,47 @@ var/const/INGEST = 2
...
+				//if a reaction only requires lists of reagents, then we want to pick the smallest list
+				smallest_number_of_reactants = INFINITY
+				smallest_reactants_list_index = 1
+
+				i = 0
+				for(var/reactant in D.required_reagents)
+					i++
+					if(islist(reactant))
+						reactant_list = reactant
+						if(smallest_reactants_list_index)
```
Comment (kacurasuo) [inline code/modules/reagents/Chemistry-Holder.dm]: since this was just defined I think it would always be true, right?

Comment (xXxEMBERGHOSTSQUADDIExXx) [reply code/modules/reagents/Chemistry-Holder.dm]: ty

**Comment by kacurasuo** [comment]
looks okay, you should test a couple of cases like reactions with many lists (like creatine) and reactions with only one reagent (slime reactions with sugars, or slime reactions with non-list reagents like blood)
just to be sure

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
sure, also while were on this topic, i dont recall, that issue with rxn collisions you mentioned in #33390, was that fixed at some point?

**Comment by kacurasuo** [comment]
> sure, also while were on this topic, i dont recall, that issue with rxn collisions you mentioned in #33390, was that fixed at some point?

yes, fixed in #33474

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
slime reaction, creatine, plant-b-gone and insecticide using solanine work so things seem to be working fine

**Comment by ShiftyRail** [comment]
seems good. I'm not sure chemical reactions are in any case a bottleneck in game performance, the main culprits usually being the light subsystem, mobs, or kudzu, but it's a net positive.

**Pull request merged**